### PR TITLE
Fix copying tasty files on incremental compilation

### PR DESCRIPTION
**Pull request by tgodzik**
Previously, we removed both generatedClassFile and rebasedClassFile from invalidations, but we only did it for rebased product files such as `.tasty`. Now we do it for both

I am not 100% sure why sometimes invalidations happen for rebased and sometimes for generated files, but that can happen according to the comment above.

What happened here is that the generated files existed, but the rebased one was in the invalidation list, so later on we would skip it.

This was causing the issues in https://github.com/scalameta/metals/pull/3706

### CODE REVIEW

**Comment by kpodsiad** [comment]
Question for Zinc newbies, what are `generatedClassFile` and `rebasedClassFile`?

**Comment by tgodzik** [comment]
> Question for Zinc newbies, what are `generatedClassFile` and `rebasedClassFile`? (generated I think is obvious, but the second one isn't)

I think the rebased one is the on where we put copy of files for the client.

**Comment by tgodzik** [comment]
Actually, now I think of it. Rebased might be a new directory and we might always be compiling to a new directory but using the artifacts from the old one. 🤔

**Comment by tgodzik** [comment]
Actually, now I think of it. Rebased might be a new directory and we might always be compiling to a new directory but using the artifacts from the old one. 🤔

**Comment by tgodzik** [comment]
> @tgodzik ELI5: what is this PR about and what are invalidations. pray Incremental compilation isn't my strong side yet :D

Ach sorry, each time anything changes in the code the files get invalidated (meaning they need recompilation) and when those get recompiled we might need to recompile some additional files etc.

File: backend/src/main/scala/bloop/BloopClassFileManager.scala [hunk 1]
```diff
@@ -180,14 +180,21 @@ final class BloopClassFileManager(
...
-          new File(rebasedClassFile.getParentFile, productName)
-        if (productAssociatedToClassFile.exists())
-          allInvalidatedExtraCompileProducts.-=(productAssociatedToClassFile)
+        new File(classFile.getParentFile, productName)
+      }
+      supportedCompileProducts.foreach { supportedProductSuffix =>
+        val generatedProductName = productFile(generatedClassFile, supportedProductSuffix)
+        val rebasedProductName = productFile(rebasedClassFile, supportedProductSuffix)
+
+        if (generatedProductName.exists() || rebasedProductName.exists()) {
```
Comment (adpi2) [inline backend/src/main/scala/bloop/BloopClassFileManager.scala:194]: I don't think it is useful to check if they exists. It is okay to remove a element from a set even if it is not there.

File: frontend/src/test/scala/bloop/BaseCompileSpec.scala [hunk 2]
```diff
@@ -140,6 +140,38 @@ abstract class BaseCompileSpec extends bloop.testing.BaseSuite {
     }
   }
 
+  test(s"compile scala 3 build incrementally") {
```
Comment (adpi2) [inline frontend/src/test/scala/bloop/BaseCompileSpec.scala:143]: Does the test fail without the fix?

**Review by adpi2 — APPROVED** [review]
While I am not sure to understand why invalidation can happen in both paths, it makes sense to remove all the products of the invalidated class files.

File: frontend/src/test/scala/bloop/BaseCompileSpec.scala (hunk 2 above)
Comment (tgodzik) [reply frontend/src/test/scala/bloop/BaseCompileSpec.scala:143]: Yes, this is based on reproduction in the Metals repository.

File: backend/src/main/scala/bloop/BloopClassFileManager.scala (hunk 1 above)
Comment (tgodzik) [reply backend/src/main/scala/bloop/BloopClassFileManager.scala:194]: I agree, but it breaks another test :sweat_smile: 

There is a test that creates a fake nir files and expects them to be removed. So if you change the target from scalaJs to jvm it should remove the old files.

**Comment by tgodzik** [comment]
I will go ahead and merge this since from my understanding this is the safest choice.

I think the main reason for this is that we change output directory on each compilation, which is why there might be different paths for invalidated files and new generated ones.
